components/autofill/src/sync/common.rs

```diff
@@ -0,0 +1,292 @@
+            }
+        };
+
+        let should_use_local = is_incoming_same || local_field == incoming_field;
+
+        if is_local_same && !is_incoming_same {
+            $merged_record.$field_name = incoming_field.clone();
+        } else if should_use_local {
+            $merged_record.$field_name = local_field.clone();
+        } else {
+            // There are conflicting differences, so we "fork" the record - we
+            // will end up giving the local one a new guid and save the remote
+            // one with its incoming ID.
+            return MergeResult::Forked {
+                forked: get_forked_record($local.clone()),
+            };
+        }
+    };
+}
+
+// And common helpers for tests (although no actual tests!)
+#[cfg(test)]
+pub(super) mod tests {
+    use super::super::*;
+    use interrupt_support::NeverInterrupts;
+    use serde::Serialize;
+    use serde_json::{json, Value};
+    use sync15::ServerTimestamp;
+
+    pub(in crate::sync) fn array_to_incoming(vals: Vec<Value>) -> Vec<(Payload, ServerTimestamp)> {
+        vals.into_iter()
+            .map(|v| {
+                (
+                    Payload::from_json(v).expect("should be a payload"),
+                    ServerTimestamp::from_millis(0),
+                )
+            })
+            .collect()
+    }
+
+    pub(in crate::sync) fn expand_test_guid(c: char) -> String {
+        c.to_string().repeat(12)
+    }
+
+    pub(in crate::sync) fn test_json_tombstone(guid_prefix: char) -> Value {
+        let t = json! {
+            {
+                "id": expand_test_guid(guid_prefix),
+                "deleted": true,
+            }
+        };
+        t
+    }
+
+    // Incoming record is identical to a local record.
+    pub(in crate::sync) fn do_test_incoming_same<
+        T: SyncRecord + std::fmt::Debug + Clone + Serialize,
+    >(
+        ri: &dyn ProcessIncomingRecordImpl<Record = T>,
+        tx: &Transaction<'_>,
+        record: T,
+    ) {
+        ri.insert_local_record(tx, record.clone())
+            .expect("insert should work");
+        let payload = Payload::from_record(record).expect("should serialize");
+        ri.stage_incoming(
+            tx,
+            vec![(payload, ServerTimestamp::from_millis(0))],
+            &NeverInterrupts,
+        )
+        .expect("stage should work");
+        let mut states = ri.fetch_incoming_states(tx).expect("fetch should work");
+        assert_eq!(states.len(), 1, "1 records == 1 state!");
+        let action =
+            crate::sync::plan_incoming(ri, tx, states.pop().unwrap()).expect("plan should work");
+        // Even though the records are identical, we still merged the metadata
+        // so treat this as an Update.
+        assert!(matches!(action, crate::sync::IncomingAction::Update { .. }));
+    }
+
+    // Incoming tombstone for an existing local record.
+    pub(in crate::sync) fn do_test_incoming_tombstone<
```

Comment:
nit: should this be limited to `crate::sync::tests`?

Reply:
Sadly, these are in `crate::sync::common::tests` and are called by, `crate::sync::addresses::tests` and `crate::sync::credit_cards::tests` - so the only shared namespace is `crate::sync`